File: backend/src/main/java/com/smartcampost/backend/model/Parcel.java
package com.smartcampost.backend.model;

import com.smartcampost.backend.model.enums.DeliveryOption;
import com.smartcampost.backend.model.enums.ParcelStatus;
import com.smartcampost.backend.model.enums.ServiceType;
import com.smartcampost.backend.model.enums.PaymentOption;
import com.smartcampost.backend.model.enums.QrStatus;
import com.smartcampost.backend.model.enums.LocationMode;
import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "parcel")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Parcel {

    @Id
    @Column(name = "parcel_id", nullable = false, columnDefinition = "BINARY(16)")
    private UUID id;

    @Column(name = "tracking_ref", nullable = false, length = 80, unique = true)
    private String trackingRef;

    // --------------------------------------------------
    //  Compatibility alias (spec wording): trackingNumber
    // --------------------------------------------------
    @Column(name = "tracking_number", length = 80)
    private String trackingNumber;

    // --------------------------------------------------
    //  🔥 QR Code Two-Step Logic (PARTIAL → FINAL)
    // --------------------------------------------------
    @Enumerated(EnumType.STRING)
    @Column(name = "qr_status", nullable = false, length = 20)
    @Builder.Default
    private QrStatus qrStatus = QrStatus.PARTIAL;

    @Column(name = "is_locked", nullable = false)
    @Builder.Default
    private boolean locked = false;

    @Column(name = "partial_qr_code", length = 500)
    private String partialQrCode;

    @Column(name = "final_qr_code", length = 500)
    private String finalQrCode;
    // --------------------------------------------------

    // --------------------------------------------------
    //  🔥 GPS Location at Creation
    // --------------------------------------------------
    @Column(name = "creation_latitude", columnDefinition = "DECIMAL(10,8)")
    private Double creationLatitude;

    @Column(name = "creation_longitude", columnDefinition = "DECIMAL(11,8)")
    private Double creationLongitude;

    @Enumerated(EnumType.STRING)
    @Column(name = "location_mode", length = 20)
    @Builder.Default
    private LocationMode locationMode = LocationMode.GPS_DEFAULT;
    // --------------------------------------------------

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "client_id",
            nullable = false,
            referencedColumnName = "client_id",
            foreignKey = @ForeignKey(name = "fk_parcel_client")
    )
    private Client client;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "sender_address_id",
            nullable = false,
            referencedColumnName = "address_id",
            foreignKey = @ForeignKey(name = "fk_parcel_sender_addr")
    )
    private Address senderAddress;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "recipient_address_id",
            nullable = false,
            referencedColumnName = "address_id",
            foreignKey = @ForeignKey(name = "fk_parcel_recipient_addr")
    )
    private Address recipientAddress;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "origin_agency_id",
            referencedColumnName = "agency_id",
            foreignKey = @ForeignKey(name = "fk_parcel_origin_agency")
    )
    private Agency originAgency; // nullable

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "destination_agency_id",
            referencedColumnName = "agency_id",
            foreignKey = @ForeignKey(name = "fk_parcel_dest_agency")
    )
    private Agency destinationAgency; // nullable

    @Column(name = "weight", nullable = false)
    private Double weight; // FLOAT en DB → Double ici

    @Column(name = "dimensions", length = 50)
    private String dimensions;

    @Column(name = "declared_value")
    private Double declaredValue; // FLOAT en DB

    @Column(name = "is_fragile", nullable = false)
    private boolean fragile;

    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false)
    private ServiceType serviceType; // STANDARD, EXPRESS

    @Enumerated(EnumType.STRING)
    @Column(name = "delivery_option", nullable = false)
    private DeliveryOption deliveryOption; // AGENCY, HOME

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private ParcelStatus status; // CREATED,...,CANCELLED

    // --------------------------------------------------
    //  🔥 SPRINT 14 — NEW FIELDS ADDED (NO DELETIONS)
    // --------------------------------------------------

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_option", nullable = false)
    private PaymentOption paymentOption; // PREPAID, COD

    @Column(name = "photo_url", length = 255)
    private String photoUrl; // optional

    @Column(name = "description_comment", length = 1000)
    private String descriptionComment; // optional comment at validation

    // --------------------------------------------------
    //  🔥 SPRINT 15 — VALIDATION FIELDS (Agent/Courier Acceptance)
    // --------------------------------------------------

    @Column(name = "validated_weight")
    private Double validatedWeight; // Weight confirmed by agent during acceptance

    @Column(name = "validated_dimensions", length = 50)
    private String validatedDimensions; // Dimensions confirmed by agent

    @Column(name = "validation_comment", length = 1000)
    private String validationComment; // Agent's notes during acceptance

    @Column(name = "description_confirmed")
    private Boolean descriptionConfirmed; // Agent confirmed description is accurate

    @Column(name = "validated_at")
    private Instant validatedAt; // Timestamp when parcel was validated/accepted

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "validated_by_staff_id",
            referencedColumnName = "staff_id",
            foreignKey = @ForeignKey(name = "fk_parcel_validated_by")
    )
    private Staff validatedBy; // Agent/Courier who validated the parcel
    // --------------------------------------------------

    @Column(
            name = "created_at",
            nullable = false,
            updatable = false,
            columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"
    )
    private Instant createdAt;

    @Column(name = "expected_delivery_at")
    private Instant expectedDeliveryAt;

    @PrePersist
    void onCreate() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        if (createdAt == null) {
            createdAt = Instant.now();
        }
        if (status == null) {
            status = ParcelStatus.CREATED;
        }
        // Keep trackingNumber in sync for clients expecting that field name
        if (trackingNumber == null && trackingRef != null) {
            trackingNumber = trackingRef;
        }
        if (!fragile) {
            fragile = false;
        }
        if (paymentOption == null) {
            paymentOption = PaymentOption.PREPAID;
        }
        // QR Code two-step logic: starts as PARTIAL
        if (qrStatus == null) {
            qrStatus = QrStatus.PARTIAL;
        }
        // Parcel starts unlocked, locked after validation
        if (!locked) {
            locked = false;
        }
        if (locationMode == null) {
            locationMode = LocationMode.GPS_DEFAULT;
        }
    }
}

File: smartcampost_mobile/lib/models/parcel.dart
class Parcel {
  final String id;
  final String? trackingNumber;
  final String? trackingRef;
  final String? status;
  final bool? locked;
  final String? qrStatus;
  final String? finalQrCode;
  final String? partialQrCode;
  final double? weight;
  final String? dimensions;
  final double? declaredValue;
  final bool? fragile;
  final String? serviceType;
  final String? deliveryOption;
  final String? paymentOption;
  final String? descriptionComment;
  final String? expectedDeliveryAt;
  final String? createdAt;
  final String? senderAddressId;
  final String? senderLabel;
  final String? senderCity;
  final String? senderRegion;
  final String? senderCountry;
  final String? recipientAddressId;
  final String? recipientLabel;
  final String? recipientCity;
  final String? recipientRegion;
  final String? recipientCountry;
  final String? clientName;
  final String? clientId;
  final String? originAgencyId;
  final String? originAgencyName;
  final String? destinationAgencyId;
  final String? destinationAgencyName;
  final String? photoUrl;
  final double? creationLatitude;
  final double? creationLongitude;
  final double? currentLatitude;
  final double? currentLongitude;
  final String? locationUpdatedAt;
  final String? locationMode;
  final double? validatedWeight;
  final String? validatedDimensions;
  final String? validationComment;
  final bool? descriptionConfirmed;
  final String? validatedAt;
  final String? validatedByStaffId;
  final String? validatedByStaffName;
  final double? lastAppliedPrice;

  Parcel({
    required this.id,
    this.trackingNumber,
    this.trackingRef,
    this.status,
    this.locked,
    this.qrStatus,
    this.finalQrCode,
    this.partialQrCode,
    this.weight,
    this.dimensions,
    this.declaredValue,
    this.fragile,
    this.serviceType,
    this.deliveryOption,
    this.paymentOption,
    this.descriptionComment,
    this.expectedDeliveryAt,
    this.createdAt,
    this.senderAddressId,
    this.senderLabel,
    this.senderCity,
    this.senderRegion,
    this.senderCountry,
    this.recipientAddressId,
    this.recipientLabel,
    this.recipientCity,
    this.recipientRegion,
    this.recipientCountry,
    this.clientName,
    this.clientId,
    this.originAgencyId,
    this.originAgencyName,
    this.destinationAgencyId,
    this.destinationAgencyName,
    this.photoUrl,
    this.creationLatitude,
    this.creationLongitude,
    this.currentLatitude,
    this.currentLongitude,
    this.locationUpdatedAt,
    this.locationMode,
    this.validatedWeight,
    this.validatedDimensions,
    this.validationComment,
    this.descriptionConfirmed,
    this.validatedAt,
    this.validatedByStaffId,
    this.validatedByStaffName,
    this.lastAppliedPrice,
  });

  String get displayRef => trackingRef ?? trackingNumber ?? id;

  factory Parcel.fromJson(Map<String, dynamic> json) => Parcel(
    id: json['id']?.toString() ?? '',
    trackingNumber: json['trackingNumber'] as String?,
    trackingRef: json['trackingRef'] as String?,
    status: json['status'] as String?,
    locked: json['locked'] as bool?,
    qrStatus: json['qrStatus'] as String?,
    finalQrCode: json['finalQrCode'] as String?,
    partialQrCode: json['partialQrCode'] as String?,
    weight: (json['weight'] as num?)?.toDouble(),
    dimensions: json['dimensions'] as String?,
    declaredValue: (json['declaredValue'] as num?)?.toDouble(),
    fragile: json['fragile'] as bool?,
    serviceType: json['serviceType'] as String?,
    deliveryOption: json['deliveryOption'] as String?,
    paymentOption: json['paymentOption'] as String?,
    descriptionComment: json['descriptionComment'] as String?,
    expectedDeliveryAt: json['expectedDeliveryAt'] as String?,
    createdAt: json['createdAt'] as String?,
    senderAddressId: json['senderAddressId']?.toString(),
    senderLabel: json['senderLabel'] as String?,
    senderCity: json['senderCity'] as String?,
    senderRegion: json['senderRegion'] as String?,
    senderCountry: json['senderCountry'] as String?,
    recipientAddressId: json['recipientAddressId']?.toString(),
    recipientLabel: json['recipientLabel'] as String?,
    recipientCity: json['recipientCity'] as String?,
    recipientRegion: json['recipientRegion'] as String?,
    recipientCountry: json['recipientCountry'] as String?,
    clientName: json['clientName'] as String?,
    clientId: json['clientId']?.toString(),
    originAgencyId: json['originAgencyId']?.toString(),
    originAgencyName: json['originAgencyName'] as String?,
    destinationAgencyId: json['destinationAgencyId']?.toString(),
    destinationAgencyName: json['destinationAgencyName'] as String?,
    photoUrl: json['photoUrl'] as String?,
    creationLatitude: (json['creationLatitude'] as num?)?.toDouble(),
    creationLongitude: (json['creationLongitude'] as num?)?.toDouble(),
    currentLatitude: (json['currentLatitude'] as num?)?.toDouble(),
    currentLongitude: (json['currentLongitude'] as num?)?.toDouble(),
    locationUpdatedAt: json['locationUpdatedAt'] as String?,
    locationMode: json['locationMode'] as String?,
    validatedWeight: (json['validatedWeight'] as num?)?.toDouble(),
    validatedDimensions: json['validatedDimensions'] as String?,
    validationComment: json['validationComment'] as String?,
    descriptionConfirmed: json['descriptionConfirmed'] as bool?,
    validatedAt: json['validatedAt'] as String?,
    validatedByStaffId: json['validatedByStaffId']?.toString(),
    validatedByStaffName: json['validatedByStaffName'] as String?,
    lastAppliedPrice: (json['lastAppliedPrice'] as num?)?.toDouble(),
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    if (trackingNumber != null) 'trackingNumber': trackingNumber,
    if (trackingRef != null) 'trackingRef': trackingRef,
    if (status != null) 'status': status,
    if (locked != null) 'locked': locked,
    if (weight != null) 'weight': weight,
    if (dimensions != null) 'dimensions': dimensions,
    if (declaredValue != null) 'declaredValue': declaredValue,
    if (fragile != null) 'fragile': fragile,
    if (serviceType != null) 'serviceType': serviceType,
    if (deliveryOption != null) 'deliveryOption': deliveryOption,
    if (paymentOption != null) 'paymentOption': paymentOption,
    if (descriptionComment != null) 'descriptionComment': descriptionComment,
    if (senderAddressId != null) 'senderAddressId': senderAddressId,
    if (recipientAddressId != null) 'recipientAddressId': recipientAddressId,
    if (clientId != null) 'clientId': clientId,
    if (originAgencyId != null) 'originAgencyId': originAgencyId,
    if (destinationAgencyId != null) 'destinationAgencyId': destinationAgencyId,
    if (photoUrl != null) 'photoUrl': photoUrl,
    if (creationLatitude != null) 'creationLatitude': creationLatitude,
    if (creationLongitude != null) 'creationLongitude': creationLongitude,
  };
}
